### PR TITLE
Improves performance of mapAsync

### PR DESCRIPTION
**Pull request by viktorklang**
Reduces allocations, indirections, shortens codepaths, reduces LoC.

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -711,78 +719,60 @@ private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Fut
...
 
-    setHandler(in, new InHandler {
       override def onPush(): Unit = {
         try {
           val future = f(grab(in))
-          val holder = new Holder[Try[Out]](NotYetThere)
+          val holder = new Holder[Out](NotYetThere, futureCB)
           buffer.enqueue(holder)
-          future.onComplete(result ⇒ futureCB.invoke(holder -> result))(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+          future.onComplete(holder)(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Look ma! No allocations!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 2]
```diff
@@ -711,78 +719,60 @@ private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Fut
...
-          }
-      })
+        } else buffer.dequeue().elem match {
+          case Success(elem) ⇒
+            push(out, elem)
+            if (todo < parallelism && !hasBeenPulled(in)) tryPull(in)
+          case Failure(ex) ⇒ pushOne()
+        }
 
-    setHandler(in, new InHandler {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Look ma! No allocations!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 3]
```diff
@@ -711,78 +719,60 @@ private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Fut
...
         }
         if (todo < parallelism) tryPull(in)
       }
-      override def onUpstreamFinish(): Unit = {
-        if (todo == 0) completeStage()
-      }
-    })
+      override def onUpstreamFinish(): Unit = if (todo == 0) completeStage()
 
-    setHandler(out, new OutHandler {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Look ma! No allocations!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 4]
```diff
@@ -711,78 +719,60 @@ private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Fut
...
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      override def toString = s"MapAsync.Logic(buffer=$buffer)"
 
-    //FIXME Put Supervision.stoppingDecider as a SupervisionStrategy on DefaultAttributes.mapAsync?
-    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+      //FIXME Put Supervision.stoppingDecider as a SupervisionStrategy on DefaultAttributes.mapAsync?
+      lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+      var buffer: BufferImpl[Holder[Out]] = _
+      val futureCB =
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Allocating the callback once sounds like a good idea.

**Comment by drewhk** [comment]
LGTM, great improvements!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 5]
```diff
@@ -791,71 +781,67 @@ private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Fut
...
+                case Failure(t)              ⇒ t
+                case Success(s) if s == null ⇒ ReactiveStreamsCompliance.elementMustNotBeNullException
+              }
+              if (decider(ex) == Supervision.Stop) failStage(ex)
+              else if (isClosed(in) && todo == 0) completeStage()
+              else if (!hasBeenPulled(in)) tryPull(in)
+          }
+        }).invoke _
 
-    setHandler(in, new InHandler {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Look ma! No allocations!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 6]
```diff
@@ -791,71 +781,67 @@ private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Fut
...
+          case NonFatal(ex) ⇒ if (decider(ex) == Supervision.Stop) failStage(ex)
         }
         if (todo < parallelism) tryPull(in)
       }
       override def onUpstreamFinish(): Unit = {
         if (todo == 0) completeStage()
       }
-    })
 
-    setHandler(out, new OutHandler {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Look ma! No allocations!

**Comment by viktorklang** [comment]
Thanks @drewhk! Let me update the commit message to be more descriptive.

**Comment by viktorklang** [comment]
@drewhk Done.

**Comment by ktoso** [comment]
LGTM, thanks a lot :)

**Comment by patriknw** [comment]
LGTM, thanks

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 1 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Fun fact, if we used FastFuture here we could speed it up even more

```scala
@Fork(2)
@State(JmhScope.Benchmark)
@BenchmarkMode(Array(Mode.Throughput))
@Warmup(iterations = 5)
@Measurement(iterations = 10)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
class FutureBenchmark {

  import akka.http.scaladsl.util.FastFuture._

  implicit val system: ActorSystem = ActorSystem()
  val dispatcher = system.dispatcher

  @TearDown
  def shutdown():Unit = {
    system.terminate()
    Await.ready(system.whenTerminated, 15.seconds)
  }

  val _normal: Future[String] = Future("hello")(dispatcher)
  val _fast: FastFuture[String] = Future("hello")(dispatcher).fast

  val fn = (s: String) => s == "hello"


  @Benchmark
  def normal_disp: Boolean = Await.result(_normal.map(fn)(dispatcher), 1.second)

  @Benchmark
  def normal_same: Boolean = Await.result(_normal.map(fn)(ExecutionContexts.sameThreadExecutionContext), 1.second)

  @Benchmark
  def fast_disp = Await.result(_fast.map(fn)(dispatcher), 1.second)

  @Benchmark
  def fast_same = Await.result(_fast.map(fn)(ExecutionContexts.sameThreadExecutionContext), 1.second)

}
```

```
[info] FutureBenchmark.fast_disp    thrpt   20  16.421 ± 3.130  ops/us
[info] FutureBenchmark.fast_same    thrpt   20  20.994 ± 0.461  ops/us
[info] FutureBenchmark.normal_disp  thrpt   20   0.080 ± 0.010  ops/us
[info] FutureBenchmark.normal_same  thrpt   20   5.998 ± 0.243  ops/us
```

// cc @viktorklang

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: I'd rather try to make the stdlib Futures faster so we can retire FastFuture :)

https://github.com/viktorklang/scala/pull/1/files

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: :+1: to that :-)
Will that get back to 2.11 though?

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: @ktoso Perhaps. More likely if it gets into 2.12 :D

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: I'd still introduce `.fast` here until we get the improvements in scala... there's a ticket, we can discuss there :)

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: you mean `future.fast.onComplete(holder)`?

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Yes, that's what I had in mind here. May happen if someone caches a completed value and calls the method (returning this, or the cached one) in mapAsync each time.

**Pull request merged**